extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/context/TransactionContext.java

```diff
@@ -0,0 +1,213 @@
+import java.util.Map;
+import java.util.concurrent.*;
+import java.util.stream.*;
+
+import javax.enterprise.context.ContextNotActiveException;
+import javax.enterprise.context.spi.Contextual;
+import javax.enterprise.context.spi.CreationalContext;
+import javax.transaction.Status;
+import javax.transaction.SystemException;
+import javax.transaction.Transaction;
+import javax.transaction.TransactionManager;
+import javax.transaction.TransactionScoped;
+import javax.transaction.TransactionSynchronizationRegistry;
+
+import com.arjuna.ats.internal.jta.transaction.arjunacore.TransactionSynchronizationRegistryImple;
+
+import io.quarkus.arc.*;
+
+/**
+ * {@link javax.enterprise.context.spi.Context} class which defines the {@link TransactionScoped} context.
+ */
+public class TransactionContext implements InjectableContext {
+    // marker object to be put as a key for SynchronizationRegistry to gather all beans created in the scope
+    private static final Object TRANSACTION_CONTEXT_MARKER = new Object();
+
+    private TransactionSynchronizationRegistry transactionSynchronizationRegistry = new TransactionSynchronizationRegistryImple();
+    private TransactionManager transactionManager = com.arjuna.ats.jta.TransactionManager.transactionManager();
+
+    @Override
+    public void destroy() {
+        if (!isActive()) {
+            return;
+        }
+        TransactionContextState contextState = (TransactionContextState) transactionSynchronizationRegistry
+                .getResource(TRANSACTION_CONTEXT_MARKER);
+        if (contextState == null) {
+            return;
+        }
+        contextState.destroy();
+    }
+
+    @Override
+    public void destroy(Contextual<?> contextual) {
+        if (!isActive()) {
+            return;
+        }
+        TransactionContextState contextState = (TransactionContextState) transactionSynchronizationRegistry
+                .getResource(TRANSACTION_CONTEXT_MARKER);
+        if (contextState == null) {
+            return;
+        }
+        contextState.remove(contextual);
+    }
+
+    @Override
+    public ContextState getState() {
+        if (!isActive()) {
+            throw new ContextNotActiveException("No active transaction on the current thread");
+        }
+
+        TransactionContextState contextState = (TransactionContextState) transactionSynchronizationRegistry
+                .getResource(TRANSACTION_CONTEXT_MARKER);
+        if (contextState == null) {
+            return new TransactionContextState<>();
+        }
+        return contextState;
+    }
+
+    @Override
+    public Class<? extends Annotation> getScope() {
+        return TransactionScoped.class;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public <T> T get(Contextual<T> contextual, CreationalContext<T> creationalContext) {
+        if (!isActive()) {
+            throw new ContextNotActiveException();
+        }
+        if (contextual == null) {
+            return null;
```

Comment:
@mkouba out of curiousity...I noticed Arc returns `null` in these cases (for req/session scoped) but Weld throws an exception. Do you think we should change it?

Reply:
+1 for `IllegalArgumentException`.

Reply:
@ochaloup https://github.com/quarkusio/quarkus/pull/4676/files brings in an exception instead of `null`, this PR should follow the same